Support to keep BuildKit state

ibalajiarun opened the pull request:
This PR adds support to keep the cache state when the builder is removed in the post-action step. For certain usecases such as CI builds, it is helpful to keep the cache and reuse it with re-created builders. For this purpose, there is a new `keep-state` optional actions flag that defaults to `false`.

To support cache reuse, this PR also allows specifying a custom name to the builder, since cache reuse is no possible with the exiting random names generated using `uuid.v4()`. For this purpose, there is a new `name` optional actions flag.

ibalajiarun commented:
Hey @crazy-max , can I get a review on this one?

ibalajiarun commented:
/cc @jedevc

crazy-max commented on src/context.ts:
```diff
@@ -37,12 +38,16 @@ export async function getInputs(): Promise<Inputs> {
-export async function getBuilderName(driver: string): Promise<string> {
-  return driver == 'docker' ? await Docker.context() : `builder-${uuid.v4()}`;
+export function getBuilderName(customName: string, driver: string): string {
+  if (customName && customName.length > 0) {
```
```suggestion
  if (customName) {
```

crazy-max commented on src/state-helper.ts:
```diff
@@ -7,6 +7,7 @@ export const builderDriver = process.env['STATE_builderDriver'] || '';
 export const containerName = process.env['STATE_containerName'] || '';
 export const certsDir = process.env['STATE_certsDir'] || '';
 export const cleanup = /true/i.test(process.env['STATE_cleanup'] || '');
+export const keepState = process.env['STATE_keepState'] || false;
```
```suggestion
export const keepState = !!process.env['STATE_keepState'];
```

crazy-max commented on action.yml:
```diff
@@ -47,6 +47,12 @@ inputs:
+    description: 'Keep state on cleanup'
+    default: 'false'
+    required: false
+  custom-name:
```
I'm not sure about naming this `custom-name`, this is confusing. Better to follow the buildx UX of the [create command](https://docs.docker.com/engine/reference/commandline/buildx_create/) and use `name`.

crazy-max reviewed (commented):
Sorry for the delay. I left some comments.

Also have you tested this change on your side? Keeping the BuildKit state will only work on persistent runners and not public ones that are ephemeral. If this is your intent then it should be documented to avoid confusion.

ibalajiarun commented:
Thanks for the review @crazy-max. I've addressed your feedback.

We have been using this change at [aptos-labs/aptos-core](https://github.com/aptos-labs/aptos-core) for over a month now with persistent self-hosted runners and have seen speed ups of up to 50% in build times.

https://github.com/aptos-labs/aptos-core/blob/751153f20f3068ed71cfe3d79ab9fe0212cc4107/.github/actions/docker-setup/action.yaml#L41-L58

ibalajiarun commented:
@crazy-max can we get this merged please?

crazy-max commented on src/context.ts:
```diff
@@ -21,12 +21,13 @@ export interface Inputs {
   return {
     version: core.getInput('version'),
-    name: await getBuilderName(core.getInput('driver') || 'docker-container'),
+    name: getBuilderName(core.getInput('name'), core.getInput('driver') || 'docker-container'),
```
`name` needs to be defined in https://github.com/docker/setup-buildx-action/blob/master/action.yml. Also needs to update README to add this new input.

crohr replied:
Being able to set a custom name would help with reusing the builder's cache across CI runs (assuming e.g. `/var/lib/docker` can be restored across runs). Would be even better if it could also check for an env var so that CI platforms (e.g. [RunsOn](https://runs-on.com)) can automatically set it without user intervention.

crazy-max reviewed (commented):
Sorry got pending reviews I forgot to submit :persevere:

It would be better if we have another PR for the `name` input being added.

zpiazza-combocurve commented:
Any movement on this?

tobiaszuercher commented:
just stumpled upon this, the `--keep-state` flag in docker buildx rm for self-hosted runners could speed up builds drastically.

what is needed to get any movement on this?

rarecrumb commented:
@ibalajiarun @crazy-max can we take another look at this? Would be great to have.

nelyodev commented:
This PR might be the missing piece of my ultimate caching pipeline quest. Any chance it will be merged in the near future?

tobiaszuercher commented:
@ibalajiarun @crazy-max is there something still open? happy to contribute

crazy-max commented on action.yml:
```diff
@@ -47,6 +47,12 @@ inputs:
     description: 'Cleanup temp files and remove builder at the end of a job'
     default: 'true'
+  keep-state:
+    description: 'Keep state on cleanup'
```
```suggestion
    description: 'Keep BuildKit state on cleanup. This is only useful on persistent self-hosted runners.'
```

crazy-max commented on action.yml:
```diff
@@ -47,6 +47,12 @@ inputs:
+    default: 'false'
+    required: false
+  name:
+    description: 'Builder name use use when creating the builder. If not specified, one will be generated. Specifying a name enables reusing a buildkit volume in persistent self-hosted runners.'
```
Last sentence not needed for this input but should be added to `keep-state` (see previous comment).

```suggestion
    description: 'Builder name set when creating the builder. If not specified, one will be generated.'
```

crazy-max commented on src/context.ts:
```diff
@@ -37,12 +38,16 @@ export async function getInputs(): Promise<Inputs> {
+  if (name) {
+    return name;
+  }
+  return driver == 'docker' ? 'default' : `builder-${uuid.v4()}`;
```
Why changing `await Docker.context()` to `default`? It might not always be default so we should keep that.

crazy-max commented on src/main.ts:
```diff
@@ -176,7 +178,7 @@ actionsToolkit.run(
         const builder = new Builder({buildx: buildx});
         if (await builder.exists(stateHelper.builderName)) {
-          const rmCmd = await buildx.getCommand(['rm', stateHelper.builderName]);
+          const rmCmd = await buildx.getCommand(['rm', stateHelper.builderName, stateHelper.keepState ? '--keep-state' : '']);
```
Although cobra should manage empty arg we should avoid this:

```suggestion
          const rmCmd = await buildx.getCommand(['rm', stateHelper.builderName, ...(stateHelper.keepState ? ['--keep-state'] : [])]);
```

crazy-max commented on src/state-helper.ts:
```diff
@@ -7,7 +7,7 @@ export const builderDriver = process.env['STATE_builderDriver'] || '';
 export const certsDir = process.env['STATE_certsDir'] || '';
 export const cleanup = /true/i.test(process.env['STATE_cleanup'] || '');
-
+export const keepState = !!process.env['STATE_keepState'];
```
Missing newline

ibalajiarun commented on src/context.ts:
```diff
@@ -42,11 +43,15 @@ export async function getInputs(): Promise<Inputs> {
 
-export async function getBuilderName(driver: string): Promise<string> {
+export async function getBuilderName(name: string, driver: string): Promise<string> {
+  if (name) {
```
That makes sense. Updated.

crohr commented:
Hi @crazy-max, any chance to get a final review on this? Thanks!

crohr commented:
This change is now used by https://github.com/runs-on/snapshot to make buildkit caches persistent across ephemeral runner jobs. Would be great to get it merged so that I can point to the official action tag :)
